staging/src/k8s.io/client-go/examples/informer/main.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Note: the example only works with the code within the same release/branch.
+package main
+
+import (
+	"flag"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	// Only required to authenticate against GKE clusters
+	// _ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+func main() {
+	var kubeconfig string
+	var master string
+
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "absolute path to the kubeconfig file")
+	flag.StringVar(&master, "master", "", "master url")
+	flag.Parse()
+
+	// BuildConfigFromFlags builds configs from a master url or a kubeconfig
+	// filepath. If neither masterUrl nor kubeconfigPath are passed, it falls back
+	// to inClusterConfig. If inClusterConfig fails, it falls back to the default
+	// config.
+	config, err := clientcmd.BuildConfigFromFlags(master, kubeconfig)
+	if err != nil {
+		glog.Fatal(err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		glog.Fatal(err)
+	}
+
+	factory := informers.NewSharedInformerFactory(clientset, time.Second*10)
+
+	informer := factory.Core().V1().Pods().Informer()
+
+	stop := make(chan struct{})
+	defer close(stop)
+
+	// the controller run starts the event processing loop
+	go informer.Run(stop)
```

Comment:
Don't call `informer.Run()`. You'll want to call `factory.Start()`, _but_ the overall order needs to be:

1. Create shared informer factory
1. Add event handler to informer you get from factory
1. Start shared informer factory
1. Wait for cache sync

With a controller, because it normally runs in its own separate goroutine, you can add the event handler and the immediately wait for the cache sync, because we know at some point later on in the main initialization code, the shared informer factory will be started at the very end of the process.